File: 04-Data-Structures/code/python/seb_ds/hash.py
"""FNV-1a 32-bit hashing, bit-for-bit identical to `typescript/src/hash.ts`.

Both implementations hash the UTF-8 byte sequence of the input rather than
any native string representation, which is what keeps them in lockstep
regardless of host language string internals.
"""

from __future__ import annotations

_OFFSET_BASIS = 0x811C9DC5
_PRIME = 0x01000193
_MASK32 = 0xFFFFFFFF


def fnv1a32(input_str: str) -> int:
    data = input_str.encode("utf-8")
    h = _OFFSET_BASIS
    for byte in data:
        h ^= byte
        h = (h * _PRIME) & _MASK32
    return h


def fnv1a32_salted(input_str: str) -> int:
    """Second, independent hash used for Bloom filter double-hashing."""
    return fnv1a32("\u0000" + input_str)


def bloom_indices(item: str, k: int, m: int) -> list[int]:
    """Deterministic k-index generator shared by BloomFilter across languages."""
    h1 = fnv1a32(item)
    h2 = fnv1a32_salted(item)
    indices = []
    for i in range(k):
        raw = h1 + i * h2
        indices.append(((raw % m) + m) % m)
    return indices

File: 04-Data-Structures/code/typescript/src/hash.ts
/**
 * FNV-1a, 32-bit variant. Bit-for-bit identical to the Python implementation
 * in `seb_ds/hash.py` for any ASCII/UTF-8 input, since both operate on the
 * UTF-8 byte sequence rather than native string encodings.
 *
 * offset basis: 0x811c9dc5, prime: 0x01000193 (the canonical FNV-1a constants).
 */
export function fnv1a32(input: string): number {
  const bytes = new TextEncoder().encode(input);
  let hash = 0x811c9dc5;
  for (let i = 0; i < bytes.length; i++) {
    hash ^= bytes[i];
    hash = Math.imul(hash, 0x01000193);
  }
  return hash >>> 0;
}

/**
 * Second, independent hash used for Bloom filter double-hashing. Salting the
 * input with a leading NUL byte before re-running FNV-1a keeps both
 * languages trivially in sync (no second algorithm to keep ported).
 */
export function fnv1a32Salted(input: string): number {
  return fnv1a32("\u0000" + input);
}

/** Deterministic k-index generator shared by BloomFilter across languages. */
export function bloomIndices(item: string, k: number, m: number): number[] {
  const h1 = fnv1a32(item);
  const h2 = fnv1a32Salted(item);
  const indices: number[] = [];
  for (let i = 0; i < k; i++) {
    const raw = h1 + i * h2;
    indices.push(((raw % m) + m) % m);
  }
  return indices;
}
